scheduler/core/components/optimizer/greedymax.py

```diff
@@ -41,97 +68,149 @@ def _allocate_time(plan: Plan, obs_time: timedelta) -> datetime:
-                if not plans[site].is_full:
-                    smax = np.max(group.group_info.scores[plans.night])
-                    maxscores.append(smax)
-                    ids.append(ii)
-                    groups.append(group)
-                    ii += 1
-                    # print(group.group.id, group.group.exec_time(), smax)
+            print(f"\nNight {plans.night + 1}")
 
-            # sort scores from high to low
-            max_score = 0.0  # maximum score in time interval
-            max_group = {}
-            jj = np.flip(np.argsort(maxscores))
-            ii = 0
-            added = False
-            # Find the group with the maximum score that will fit in the remaining time
-            while not added and ii < len(jj):
-                max_score = maxscores[jj[ii]]
-                max_group = groups[jj[ii]]
-                # for obs in group.group.observations():
-                #     print(f'\t {obs.id} {obs.exec_time()}')
-                #     for atom in obs.sequence:
-                #         print(f'\t\t {atom.id} {atom.exec_time}')
-                if max_score > 0.0:
-                    # Try to add group
-                    added = self.add(max_group, plans)
-                    if added:
-                        # TODO: All observations in the group are being inserted so the whole group
-                        # can be removed from the active group list
-                        print(f'{max_group.group.unique_id()} with max score {max_score} added.')
-                        self.groups.remove(max_group)
-                    else:
-                        print(f'{max_group.group.unique_id()} with max score {max_score} not added.')
-                else:
-                    # Finished, nothing more to schedule
-                    for plan in plans:
-                        plan.is_full = True
-                    break
-                ii += 1
-            if not added and ii == len(jj):
+            # Find the group with the max score in an open interval
+            max_score, max_group, max_interval = self._find_max_group(plans)
+
+            # If something found, add it to the timeline and plan
+            if max_interval is not None:
+                added = self.add(max_group, plans, max_interval)
+                if added:
+                    print(f'{max_group.group.unique_id()} with max score {max_score} added.')
+                    self.groups.remove(max_group)  # should really only do this if all time used (not split)
+            else:
                 # Nothing remaining can be scheduled
                 for plan in plans:
                     plan.is_full = True
+                for timeline in self.timelines[plans.night]:
+                    timeline.is_full = True
 
-            print('')
-
-    def add(self, group: GroupData, plans: Plans) -> bool:
+    def add(self, group: GroupData, plans: Plans, interval) -> bool:
```

Comment:
I suggest we move `Interval` from `Timeline` to a common class, since with this change, it would be nice to have it available to `BaseOptimizer` even though it's just a type alias.

I'm concerned that keeping it where it is will introduce a circular dependency. I suggest we put it in the `optimizer` package's `__init__.py` file. I'll go ahead and do that.

Reply:
Note that the new signature of this method will be:

```python
    def add(self, group: GroupData, plans: Plans, interval: Optional[Interval]) -> bool:
```